bpo-36389: Add gc.enable_object_debugger()

vstinner opened the pull request:
New "object debugger" which checks frequently if all Python object tracked
by the garbage collector are consistent: gc.enable_object_debugger()
and gc.disable_object_debugger().

* Add new _PyObject_CheckConsistency() function
* _PyUnicode_CheckConsistency() and _PyDict_CheckConsistency() are
  now exposed in the internal API. _PyDict_CheckConsistency()
  parameter type becomes PyObject*.
* Add more checks to _PyType_CheckConsistency().

<!-- issue-number: [bpo-36389](https://bugs.python.org/issue36389) -->
https://bugs.python.org/issue36389
<!-- /issue-number -->

vstinner committed:
bpo-36389: Add gc.enable_object_debugger()

New "object debugger" which checks frequently if all Python objects tracked
by the garbage collector are consistent: gc.enable_object_debugger()
and gc.disable_object_debugger().

* Py_FatalError() and _PyObject_AssertFailed() now disable the GC
  object debugger to prevent reentrant calls.
* Fix _PyObject_Dump() for ob_type=NULL

csabella commented on Doc/whatsnew/3.8.rst:
```diff
@@ -182,6 +182,11 @@ gc
 indicating a generation to get objects from. Contributed in
 :issue:`36016` by Pablo Galindo.
 
+New "object debugger" which checks frequently if all Python object tracked by
```
s/object/objects

csabella commented on Doc/library/gc.rst:
```diff
@@ -263,6 +263,28 @@ values but should not rebind them):
 
 
+The :mod:`gc` module provides a "object debugger" which checks frequently if
+all Python object tracked by the garbage collector are consistenct.
```
```suggestion
The :mod:`gc` module provides an "object debugger" which checks frequently if
all Python objects tracked by the garbage collector are consistent.
```

csabella commented on Doc/library/gc.rst:
```diff
@@ -263,6 +263,28 @@ values but should not rebind them):
+   memory allocation or deallocation made by the garbage collector.
+
+   Low *threshold* can have a significant negative impact on Python
+   performances, but should detect earlier inconsistent Python objects.
```
s/performances/performance

vstinner committed:
Add multiple thresholds

Disable debugger in PyObject_GC_Del() and _PyObject_GC_Resize(): the
debugger is just too slow.

serhiy-storchaka commented on Modules/gcmodule.c:
```diff
@@ -1240,6 +1243,107 @@ collect_generations(void)
+{
+    struct _gc_object_debugger *debugger = &_PyRuntime.gc.object_debugger;
+    debugger->enabled = 0;
+    for (int i=0; i < NUM_GENERATIONS; i++) {
```
```suggestion
    for (int i = 0; i < NUM_GENERATIONS; i++) {
```

serhiy-storchaka commented on Modules/gcmodule.c:
```diff
@@ -1240,6 +1243,107 @@ collect_generations(void)
+    GC_OBJECT_ASSERT(op, type != NULL);
+    GC_OBJECT_ASSERT(op, !_PyObject_IsFreed((PyObject *)type));
+
+#undef ASSERT
```
What is ASSERT?

serhiy-storchaka commented on Modules/gcmodule.c:
```diff
@@ -1683,6 +1787,70 @@ gc_get_freeze_count_impl(PyObject *module)
 
 
+/*[clinic input]
+gc.enable_object_debugger as gc_py_enable_object_debugger -> NoneType
```
I would not use `-> NoneType`. It does not add clarity. Actually `return Py_None` without increfing looks suspicious. It is clearer to use Py_RETURN_NONE.

serhiy-storchaka commented on Doc/library/gc.rst:
```diff
@@ -263,6 +263,39 @@ values but should not rebind them):
+
+This debugger aims to debug bugs in C extensions.
+
+.. function:: enable_object_debugger(threshold)
```
The signature does not match implementation.

serhiy-storchaka commented on Include/internal/pycore_pymem.h:
```diff
@@ -111,6 +111,15 @@ struct gc_generation_stats {
+        int threshold; /* collection threshold */
+        int count; /* count of allocations or collections of younger
+                      generations */
+    } generations[NUM_GENERATIONS];
```
Sorry, I have not found where multiple generations are used?
